### PR TITLE
Components

### PR DESCRIPTION
**Pull request by L0neGamer**
Adding components support, creating a bunch of helper functions and types, updating to version 8, making it so things actually work, adding some examples

### CODE REVIEW

**Review by aquarial — COMMENTED** [review]
Wow! This is a massive change for some ridiculous complexity from discord docs API. 

TBH I'm not quite sure what to do with this. There are parts that are good changes, and parts that are so ugly that I'm sad with the docs. Hmm

File: src/Discord/Internal/Types/Prelude.hs [hunk 1]
```diff
@@ -71,3 +78,41 @@ epochTime :: UTCTime
...
+
+instance ToJSON InteractionType where
+  toJSON = toJSON . fromEnum
+
+instance FromJSON InteractionType where
+  parseJSON = withScientific "InteractionType" (return . toEnum . round)
+
+class Internals a b where
+  toInternal :: a -> b
+  fromInternal :: b -> Maybe a
```
[the remark this answers is not in the document]
Comment (aquarial) [reply src/Discord/Internal/Types/Prelude.hs:114]: In other places I've used the pattern of a function like `createEmbed :: CreateEmbed -> Embed`, where `CreateEmbed` has the user controlled aspects, and `Embed` has the actual message.

[the remark this answers is not in the document]
Comment (L0neGamer) [reply src/Discord/Internal/Types/Prelude.hs:114]: I'm aware that my conversions may be a bit delicate, hence my want to make sure that people could use the internal structures if they felt bold enough.

[the remark this answers is not in the document]
Comment (L0neGamer) [reply src/Discord/Internal/Types/Prelude.hs:114]: but such an isolation should be possible of course.

File: src/Discord/Internal/Types/ApplicationCommands.hs [hunk 2]
```diff
@@ -1,36 +1,230 @@
...
+  toInternal ApplicationCommandOptionValueInteger {..} = InternalApplicationCommandOption ApplicationCommandOptionTypeInteger applicationCommandOptionValueName applicationCommandOptionValueDescription applicationCommandOptionValueRequired (((StringNumberValueInteger <$>) <$>) <$> applicationCommandOptionValueIntegerChoices) Nothing Nothing (fromInteger <$> applicationCommandOptionValueIntegerMinVal) (fromInteger <$> applicationCommandOptionValueIntegerMaxVal) applicationCommandOptionValueAutocomplete
+  toInternal ApplicationCommandOptionValueString {..} = InternalApplicationCommandOption ApplicationCommandOptionTypeInteger applicationCommandOptionValueName applicationCommandOptionValueDescription applicationCommandOptionValueRequired (((StringNumberValueString <$>) <$>) <$> applicationCommandOptionValueStringChoices) Nothing Nothing Nothing Nothing applicationCommandOptionValueAutocomplete
```
[the remark this answers is not in the document]
Comment (L0neGamer) [reply src/Discord/Internal/Types/ApplicationCommands.hs:153]: oh, that's a thing - StringNumberValue now stores all three types so I can take advantage of the parsers catching out invalid integers.

**Pull request closed without merging**